### ISSUE TITLE
Linux mouse issues

### ISSUE THREAD

**Issue by OBattler** [86Box/86Box#4239]
Creating this unified issue to track Linux mouse issues.

Currently known:
- https://github.com/86Box/86Box/issues/3937;
- https://github.com/86Box/86Box/issues/2331.

**Comment by rakslice** [comment]
Based on the observed behaviour, the 86box's Linux mouse capture approach is to, on mouse movements, overwrite the mouse cursor position to fixed positions on the screen.

If the cursor position turns out to be inside the part of the 86box window displaying the emulated video, then everything appears to work: middle click exits mouse capture properly without sending a click anywhere else and other clicks go to the emulated system. But if the cursor position is not inside that part of the 86box window, then you're just clicking there, and whatever other window at that position gets focused or UI element there gets clicks.

And tl;dr the way 86box chooses those positions is broken in ways that are too complicated to explain, but test with your multiple monitors in different scalings, with the 86box window in different positions, and with the emulated system in different video modes, and you should easily come up with lots of different broken cases to fix.

And I'm just testing in Ubuntu 22.04, previous LTS, with nvidia driver 470.256.02 X11, with a conventional mouse. Who knows what happens in wayland or with a trackpad, touchscreen, other absolute pointing device, remote control with x2x or similar, etc.

**Comment by JuananBow** [comment]
This might not represent the magnitude of the problem and could be more like a niche situation, but I had been doing some testing using Steam Deck running the latest version of Holo OS (3.5.19). 

In the so called "Gaming mode" the mouse support totally got broken somewhere between the build 4311 (3.11) and the build 5000 (4.00). To be precise, it got broken on **build 4410**; a build apparently generated to address VNC mouse capture related problems specifically. According to the changelog: 

_xinput2_mouse: Treat XTEST as an absolute pointer device to fix VNC mouse injection ([commit: e9a3266](https://github.com/86Box/86Box/commit/e9a326622fc1e9cbc7d013d3231b77b6a81ddc32))_ 

On the other side, in the "Desktop mode", capturing the mouse from a VNC window does not seem to work at all. No matter if I am on a "pre-4410" build or a "post-4410". Although the mouse works if move it using the physical inputs of the console (pad or physical mouse connected via USB)

**Comment by rakslice** [comment]
This just plain wrong, right, like the comment doesn't match the condition? https://github.com/86Box/86Box/blob/882aabf95d198b3df2fc781aa0b1f6b5b38b9f91/src/qt/xinput2_mouse.cpp#L179

**Comment by OBattler** [comment]
> This just plain wrong, right, like the comment doesn't match the condition?
> 
> https://github.com/86Box/86Box/blob/882aabf95d198b3df2fc781aa0b1f6b5b38b9f91/src/qt/xinput2_mouse.cpp#L179

It does much - in some cases (I believe some virtualizers), XTEST claims to be in relative mode while in fact using absolute coordinates, hence the workaround.

**Comment by richardg867** [comment]
XTEST is used by some VNC servers for input injection, specifically the licensed UltraVNC formerly used by Raspberry Pi OS, while other servers like TigerVNC create their own devices. In both cases, the device reports that it has relative axes like a mouse, but produces absolute coordinates through XInput2, while cursor movement itself is done through a different method.

The line 179 check only applies to these "fake relative" devices if one is detected; see also the line 168 check which is where actual relative pointers are handled.

While build 4410 did break Steam Deck touchpad support, build 4633 fixed it by disabling the XTEST workaround if a Steam Deck controller is present. If that workaround is no longer functional in recent versions of SteamOS, I'd like the output of `xinput list` on it.

The pointer centering mechanism has already proven to be an issue with X11 forwarding (VcXsrv on Windows is yet another example of a fake relative pointer!) where the warp calls (and also the pointer device queries but that's easily cacheable) take *forever* even on a local network. Better ideas for how to lock the pointer without locking movement entirely are always appreciated.

**Comment by rakslice** [comment]
Okay, I had an opportunity to look into it.

The code tries to reposition the cursor to a fixed point (the center of the window) after you move the mouse -- but the cursor is not just going to one incorrect fixed position but oscillating between multiple incorrect fixed positions because the code is setting the cursor position in two different ways:

- Using Qt's `QCursor:setPos()`, in `RendererStack::mouseMoveEvent()` in `src/qt/qt_rendererstack.cpp`
- Using xlib's `XWarpPointer()` in `xinput2_proc()` in `src/qt/xinput2_mouse.cpp`

and they're both malfunctioning in different ways because of different problems.

I'm running Qt 5.15:
```
$ dpkg -l | grep qtbase
ii  qtbase5-dev:amd64                             5.15.3+dfsg-2ubuntu0.2                            amd64        Qt 5 base development files
```
In this version:

- In `QWindow::mapToGlobal()`, the position within the screen component of the returned global coordinates does not reflect the device pixel ratio (whether that's due to the desktop environment scaling setting or `QT_SCALE_FACTOR` etc.)
- `QCursor::setPos()` is just broken; on the second monitor it ignores the x coordinate and always puts the cursor at the right edge of the screen, even if you just `QCursor::setPos(QCursor::pos())` with the cursor on the second monitor.

As a cheesy workaround for these, for the `xinput2_mouse.cpp` version of the cursor warping, I did the math separately to fix the value from `mapToGlobal()`; and then to work around the broken `QCursor::setPos()` I just factored the other version out of `xinput2_proc()` and replaced the broken code in `RendererStack` with a call to it.

https://github.com/86Box/86Box/compare/master...rakslice:86Box:mouse_capture_linux_workaround

This fixed the problems with the mouse cursor going places unexpectedly and with the buttons not working properly due to that.

Obviously that's just a quick workaround and needs to be reworked to allow for systems that use qt but not xinput2 and whatever Qt versions have working `mapToGlobal()` whose coordinates should not be messed with.

**Comment by alexolivan** [comment]
Hi!

Gave a try to 86Box (64bit appImage) on Debian 11/Bullseye with Mate Desktop  (just x11, no wayland around) and I stumbled with mouse issues that render the whole emulator simply unusable.

Upon clicking the screen, the pointer disapears. Either freeing mouse capture and/or moving the mouse around, it finally makes the pointer to show, but stuck on the right border of the emulator window, like if the virtual mouse had its horizontal axis 'locked'... you can move the cursor up and down along  the right side border, but completelly unresponsive to any left / right input, so you just see part of the pointer moving up and down.

Sometimes, all of the sudden (I do not find a clear pattern to replicate this, it simply happens somehow moving the mouse around for a while), the cursor seems to stop being 'horizontally locked', like breaking free, and moving into the useable area of the emulator window, and then, from then on, it works absolutelly normal! ... until freeing host mouse (ctrl + end) and trying again to get the mouse on the emulator, which locks up the pointer to the right border.

I initially though on a Guest OS mouse driver issue, but after being able to use the mouse normally at some point, that made me think it is rather an emulator issue handling the mouse/cursor positioning.

Nice project! It was very exciting to see again my teenage AMD 386DX 40 memory count on the BIOS boot!
Hope its fixed someday.

Best regards.
Cheers.

**Comment by ColetteDiskette** [comment]
I'm having the mouse shooting to the right issue, myself. Every time I put focus back on the window, it goes to the far right and after a certain period of time it sometimes comes back "in-bounds," and other times I have to quickly move the mouse back and forth to force it back in-bounds.

I'm on 86Box 4.2.1 on Arch Linux running AwesomeWM (which is under X11.)

**Comment by the-burrito-triangle** [comment]
@TheBombSquad

Using the latest Linux AppImage build (b6327), the cursor still gets stuck in the lower right hand side of the window when running a Win98se VM.

HOWEVER, if I check the "HiDPI scaling" box under "View: Specify dimensions" AND set "Window scale factor" to "2x", then it works fine... Any other setting and the cursor gets stuck.

This is with Fedora 39 Workstation (GNOME DE / Wayland) and a 4k monitor with 200% scaling.

Would it be possible to port the DOSBox-X mouse code over? Jon has the scroll wheel working in Windows 98 and the overall cursor experience feels light years ahead of 86Box. (But to be fair, 86Box has double the disk performance of DBX: 6.4MiB/s read/write vs 2.8/3.6MiB/s read/write with CrystalDiskMark v2.2 using a 1GiB 4KiB aligned FAT32 vhd Win98se guest.)

**Comment by the-burrito-triangle** [comment]
So...I spoke too soon. If I manually edit the config file it actually seems to work! If I change setting (HiDPI or Window scale factor) while running the VM the mouse gets stuck. So there is a bug somewhere, but manually editing the config seems to work fine if the HiDPI box is checked. I can't seem to find that setting in the cfg file to test manually editing it.

**Comment by the-burrito-triangle** [comment]
So, testing with Diablo I and it's expansion Hellfire, the cursor will still occasionally jump to the far right bottom corner but not get stuck. This is with 2x Window scale factor and HiDPI scaling checked (which ends up being 4x scaling overall due to HiDPI matching the desktop's 200% scaling).

So yeah... the cursor is still wonky.

I'm impressed that 86Box passes the dxdiag D3D7-9 tests with a PCI Voodoo 3. The funny thing is, the D3D7 and D3D8 tests (a 3D spinning cube) had significantly less CPU load than the 2D DDraw tests (a white bouncing square reminiscent of pong)! Lol. The DDraw performance is pretty bad with obvious tearing and maxed out CPU loading. DOSBox-X is better here, but only has DDraw and Glide support--no D3D. If the mouse input and 2D acceleration were better 86Box would be amazing for Win9x emulation.

**Comment by OBattler** [comment]
6237 did not yet have TheBombSquad's fixes, those were introduced in 62**9**7. I recomend updating to the latest build and testing to see if it works now.

**Comment by OBattler** [comment]
Wait, you weore 6327, not 6237. My bad. Then the fixes should already be in. His fixes only apply to XInput, though, so try switching to that.

**Comment by Dizzy611** [comment]
Just tested the Tandy 1000 SL/2 Deskmate Serial Mouse issue (issue #1 in the list here) and it's still present. However it's worth noting that when recently testing SimCity on an emulated IBM PC (1982) I noticed no such mouse issues. Testing SimCity on a Tandy 1000 SL/2 however, I do notice these same mouse issues.

TL;DR, I'm not actually sure this *is* a Linux mouse issue, as much as a Tandy 1000 SL/2 specific serial mouse issue.

**Comment by OBattler** [comment]
I suspect it's the latter. Because I don't see how a Linux mouse issue would only affect one single machine.

**Comment by OBattler** [comment]
Closing this because it's being united with this issue: https://github.com/86Box/86Box/issues/4824 for overall Linux-specific issues since there's few enough of them that two separate issues are now overkill.
